### PR TITLE
Compile error propagation

### PR DESCRIPTION
**Pull request by vext01**
See commits. Main work is the second.

### CODE REVIEW

File: ykrt/src/compile/mod.rs [hunk 1]
```diff
@@ -28,6 +27,50 @@ pub(crate) mod jitc_llvm;
...
+    Temporary(TemporaryErrorKind),
+}
+
+impl CompilationError {
+    pub fn new_jitir_index_overflow(which: String) -> Self {
+        Self::Temporary(TemporaryErrorKind::JITIRIndexOverflow(which))
+    }
+}
+
+pub type CompilationResult<T> = Result<T, CompilationError>;
```
Comment (ltratt) [inline ykrt/src/compile/mod.rs:45]: I must admit I really dislike `type` aliases used in this way: they save very little typing (in this case 6 characters) but now I have to know what they alias. So I find they make the code worse. I appreciate this is a personal quirk of mine so feel free to ignore if you disagree.

CompilationResult<T>
Result<T, CompilationError>

File: ykrt/src/compile/mod.rs [hunk 2]
```diff
@@ -28,6 +27,50 @@ pub(crate) mod jitc_llvm;
 #[cfg(jitc_yk)]
 pub mod jitc_yk;
 
+/// A failure to compile a trace.
+#[derive(Debug, thiserror::Error)]
+pub enum CompilationError {
+    #[error("Unrecoverable error: {0}")]
+    Unrecoverable(String),
+    #[error("Temporary error: {0}")]
+    Temporary(TemporaryErrorKind),
```
Comment (ltratt) [inline ykrt/src/compile/mod.rs:36]: Do we gain anything useful from `TemporaryErrorKind`? My guess is that, at least at the moment, we will not find its contents valuable other than printing them out for debugging. Perhaps in most cases we even want a `temporary_compilation_error!()` macro which expands to `CompilationError::Temporary(format!("Line {}", line!())` or similar? It cuts down on us having to type stuff but means that when we're debugging we'll know *which* temporary error caused a problem.

File: ykrt/src/compile/mod.rs (hunk 1 above)
Comment (vext01) [reply ykrt/src/compile/mod.rs:45]: I think this is one of the few cases where a type alias is acceptable. It's an idiom in use in the rust std lib, so I'm actually OK with it.

@ptersilie do you have any preference?

Comment (ltratt) [reply ykrt/src/compile/mod.rs:45]: Yes it is used in the standard library, but it really makes a mess there too IMHO.

File: ykrt/src/compile/mod.rs (hunk 2 above)
Comment (vext01) [reply ykrt/src/compile/mod.rs:36]: I think you are right. We don't actually care what's inside. I just copied precedence from [elsewhere](https://github.com/ykjit/yk/blob/7dc01b9ccc79bf629a2136cf545bc6db35c3ffd0/hwtracer/src/errors.rs#L15).

Maybe that should be updated at some point too (but not now).

Comment (vext01) [reply ykrt/src/compile/mod.rs]: Fixed in https://github.com/ykjit/yk/pull/956/commits/ab849eb9d74e139a62c5466dca6f2c283162f1e5

File: ykrt/src/compile/mod.rs (hunk 1 above)
Comment (vext01) [reply ykrt/src/compile/mod.rs]: Changed in https://github.com/ykjit/yk/pull/956/commits/b169db4fbe9dcbb8e3b3bad5f4eeffd0e0b3ddd1

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Comment by ltratt** [comment]
@vext01 Those are genuine (if minor) errors, which will need fixing.

**Comment by vext01** [comment]
I guess this is the way to fix it? https://github.com/ykjit/yk/pull/956/commits/55ded243bb6bc95a455e01cc92a0f9a9d8926dcc

**Comment by ltratt** [comment]
> I guess this is the way to fix it?

Nope :) You'll need to conditionally compile the whole match arm.

**Comment by vext01** [comment]
Yeah, i was hoping not to...

**Comment by vext01** [comment]
Someone meh fix in https://github.com/ykjit/yk/pull/956/commits/da2a6133ff707b501a823dce2fad87499fb70c55

**Comment by ltratt** [comment]
Hmm, I stand corrected. https://github.com/ykjit/yk/commit/55ded243bb6bc95a455e01cc92a0f9a9d8926dcc is less evil. Please revert da2a613.

**Comment by vext01** [comment]
OK to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Pull request merged**